pkg-manager/core/src/install/index.ts

```diff
@@ -566,14 +572,31 @@ export async function mutateModules (
         optionalDependencies,
         updateWorkspaceDependencies: project.update,
         preferredSpecs,
-        saveCatalogName: opts.saveCatalogName,
+        saveCatalogName,
         overrides: opts.overrides,
         defaultCatalog: opts.catalogs?.default,
       })
+
       projectsToInstall.push({
         pruneDirectDependencies: false,
         ...project,
-        wantedDependencies: wantedDeps.map(wantedDep => ({ ...wantedDep, isNew: !currentBareSpecifiers[wantedDep.alias], updateSpec: true, nodeExecPath: opts.nodeExecPath })),
+        wantedDependencies: wantedDeps.map(wantedDep => {
+          if (saveCatalogName && !wantedDep.bareSpecifier.includes(':')) {
+            const catalogDep = opts.catalogs?.[saveCatalogName]?.[wantedDep.alias]
```

Comment:
Can the `resolveFromCatalog` function and the corresponding matcher from `@pnpm/catalogs.resolver` be used here? Centralizing the logic there makes it easier to change resolution logic in the future.